### PR TITLE
Splits context menu into separate component

### PR DESCRIPTION
**Pull request by hoonji**
## Motivation for features / changes
Splits DataTableComponent's context menu functionality into its own component to be compatible with future updates that will enable dynamic custom modal creation (for alleviating [stacking context issues](https://github.com/tensorflow/tensorboard/pull/6737#discussion_r1469588008))

## Technical description of changes
- No new functionality - simply moves the context menu related template and logic to a new ContextMenuComponent
- Parameterizes some context menu tests for legibility.

## Detailed steps to verify changes work correctly (as executed by you)
- Tested manually that all context menu features work as before

### CODE REVIEW

File: tensorboard/webapp/widgets/data_table/context_menu_component.ng.html
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/context_menu_component.ng.html:1]: copied from data_table_component.ng.html

File: tensorboard/webapp/widgets/data_table/context_menu_component.scss
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/context_menu_component.scss:1]: Copied from data_table_component.scss

File: tensorboard/webapp/widgets/data_table/context_menu_component.ts
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/context_menu_component.ts:1]: Copied from data_table_component.ts

File: tensorboard/webapp/widgets/data_table/context_menu_test.ts
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/context_menu_test.ts:1]: Copied from data_table_test.ts

File: tensorboard/webapp/widgets/data_table/data_table_component.ng.html [hunk 1]
```diff
@@ -104,7 +56,7 @@
       <button
         mat-icon-button
         class="add-column-btn"
-        (click)="openColumnSelector($event)"
+        (click)="openColumnSelector({event: $event})"
```
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/data_table_component.ng.html:59]: To allow ContextMenuComponent to propagate openColumnSelector parameters through its output, we need to make openColumnSelector's parameters an object as an Event can only be an object.

File: tensorboard/webapp/widgets/data_table/data_table_component.ts [hunk 2]
```diff
@@ -363,8 +328,8 @@ export class DataTableComponent implements OnDestroy, AfterContentInit {
     });
   }
 
-  openFilterMenu(event: MouseEvent, header: ColumnHeader) {
```
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/data_table_component.ts:366]: header is exactly equivalent to this.contextMenuHeader

File: tensorboard/webapp/widgets/filter_input/filter_input_component.scss [hunk 3]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 :host {
   display: flex;
   font-size: 13px;
+  min-width: 210px;
```
Comment (hoonji) [inline tensorboard/webapp/widgets/filter_input/filter_input_component.scss:20]: Fixes discrete filter placeholder text from being cut off: 
6tXMnVtA3ELS7c2

File: tensorboard/webapp/widgets/data_table/data_table_component.ts [hunk 4]
```diff
@@ -302,59 +308,18 @@ export class DataTableComponent implements OnDestroy, AfterContentInit {
...
-      return index;
-    }
-    if (this.insertColumnTo === Side.RIGHT) {
-      return Math.min(index + 1, this.headers.length);
-    }
-
-    return index;
-  }
-
   onColumnAdded(header: ColumnHeader) {
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/data_table_component.ts:323]: I played with your changes. "Insert Column Left" no longer works as expected. It inserts to the right instead.

**Review by bmd3k — APPROVED** [review]

File: tensorboard/webapp/widgets/data_table/data_table_component.ts (hunk 4 above)
Comment (hoonji) [reply tensorboard/webapp/widgets/data_table/data_table_component.ts:319]: Oh thanks for catching this! To clarify, the quoted "getInsertIndex()" code should be irrelevant as it's unused, but the other changes introduced a race condition where insertColumnTo is reset to undefined after it is first properly set on column open.

Added a temporary fix (don't reset insertColumnTo to undefined - this is actually unnecessary) and included a simple regression test. Note that this class of problem should be irrelevant in the next PR that creates modals dynamically.

**Pull request merged**